Adds Hilbert driven standard basis computation

RafaelDavidMohr opened the pull request:
This adds the function `_groebner_basis_with_hilbert` and adds calling it as an `:algorithm` option to `standard_basis` and `groebner_basis`. 

This function computes the hilbert series of an ideal (probabilistically over a randomly chosen prime if the characteristic of the underlying ring is zero) and passes it to `std_hilbert`. 

One thing I am unsure about: Should the user be able to pass the weights into `standard_basis` and `groebner_basis` explicitly or should they be extracted out of the underlying polynomial ring?

RafaelDavidMohr committed:
groebner with hilbert prototype

RafaelDavidMohr committed:
prototype for std with hilbert function

RafaelDavidMohr committed:
pick a random prime in standard_basis_hilbert

RafaelDavidMohr committed:
simplifies standard_with_hilbert

RafaelDavidMohr committed:
deletes unused variable

RafaelDavidMohr committed:
adds weights to homogeneity check

RafaelDavidMohr committed:
puts `_groebner_basis_with_hilbert` into `standard_basis`

RafaelDavidMohr committed:
make sure we don't compute the same GB twice

RafaelDavidMohr committed:
fix typo in error message

RafaelDavidMohr committed:
adds `_groebner_basis_with_hilbert` to `groebner_basis`

RafaelDavidMohr committed:
adds call to `Singular.std_hilbert` and short example

thofma commented:
The docstring says `Int`, but the implementation says `Int32`. I would suggest to use `Int` everywhere one the julia side. Since there is already an explicit conversion before calling Singular, this should be fine.

If the Hilbert series is wrong, will the Gröbner/standard basis still be correct?

ederc commented:
> If the Hilbert series is wrong, will the Gröbner/standard basis still be correct?

The process the Hilbert series is constructed is semicontinuous, so even if the Hilbert series is not correct (due to a bad prime for example) the computation might only be less efficient (more S-polys to be checked w.r.t. the destination ordering), but the resulting basis will still be correct.

ederc commented on src/Rings/groebner.jl:
```diff
@@ -137,7 +141,7 @@ negdegrevlex([x, y])
 """
 function standard_basis(I::MPolyIdeal; ordering::MonomialOrdering = default_ordering(base_ring(I)),
-	complete_reduction::Bool = false, algorithm::Symbol = :buchberger)
+	complete_reduction::Bool = false, algorithm::Symbol = :buchberger, weights::Vector{E} = ones(Int32, ngens(base_ring(I)))) where {E <: Integer}
```
We do not add `weights` to `standard_basis()`, same holds for `groebner_basis()`. We need to compute this data internally.

ederc commented on src/Rings/groebner.jl:
```diff
@@ -249,9 +258,9 @@ julia> leading_coefficient(G[8])
 """
 function groebner_basis(I::MPolyIdeal; ordering::MonomialOrdering = default_ordering(base_ring(I)), complete_reduction::Bool=false,
-	algorithm::Symbol = :buchberger)
+	algorithm::Symbol = :buchberger, weights::Vector{E} = ones(Int32, ngens(base_ring(I)))) where {E <: Integer}
```
See above.

ederc commented on src/Rings/groebner.jl:
```diff
@@ -1221,3 +1232,104 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
+
+# TODO: does the target ordering have to be global?
+@doc Markdown.doc"""
+    _groebner_basis_with_hilbert(I::MPolyIdeal,
```
Make this an exported function called `groebner_basis_hilbert_driven()`, we will later on adjust calls to `f4` and `fglm` to this naming convention. Here you correctly have the `weights` as parameters, but we do not have them for the general `..._basis()` calls.

ederc commented on src/Rings/groebner.jl:
```diff
@@ -1221,3 +1232,104 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
+      while true
+        p = rand(2^15:2^16)
+        if isprime(p)
+          break
```
```suggestion
    r = rand(2^15:2^16)
    p = next_prime(r)
```

Maybe that's more efficient?

ederc commented on src/Rings/groebner.jl:
```diff
@@ -1221,3 +1232,104 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
 end
+
+
+# TODO: does the target ordering have to be global?
```
Yes, only global orderings are supported.

RafaelDavidMohr committed:
simpler prime finding

RafaelDavidMohr committed:
renames groebner hilbert function and exports it

RafaelDavidMohr committed:
adds helper function to extract integer weights from graded ring

RafaelDavidMohr committed:
updates documentation with new function name

RafaelDavidMohr committed:
removes weights parameter from `standard_basis` and `groebner_basis`

RafaelDavidMohr committed:
removes TODO comment

RafaelDavidMohr committed:
fixes error message in `groebner_basis_hilbert_driven`

RafaelDavidMohr committed:
adds error if destination ordering is a general matrix ordering

RafaelDavidMohr commented:
Adjusted the code based on your reviews, see the last few commits. If I understood correctly, the `std_hilbert` function in Singular does not work with general matrix orderings as well, so I added a corresponding error message.

thofma commented on src/Rings/groebner.jl:
```diff
@@ -1221,3 +1231,112 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
+        I_mod_p_gens = [ModP(f) for f in gens(I)]
+        G = f4(ideal(ModP, I_mod_p_gens))
+        break
+      catch
```
What is the possible error? The error should be checked and rethrown if it does not fit.

RafaelDavidMohr replied:
This error is caused if one of the coefficients of the input polynomials is divisible by our prime `p`. I changed the code to handle specifically only this error (see https://github.com/oscar-system/Oscar.jl/pull/1821/commits/89e6371abc0aabc6da63938f43b2f637fc9c3df6), I don't know if there is a better way to handle this.

thofma replied:
Looks good, thanks.

RafaelDavidMohr committed:
fixes a syntax error

RafaelDavidMohr committed:
better coercion error handling

RafaelDavidMohr committed:
change throw to rethrow

RafaelDavidMohr committed:
helper function to lift a monomial order to the homogenization

RafaelDavidMohr committed:
homogenizes in `standard_basis` if needed

RafaelDavidMohr commented:
I've modified `standard_basis` to homogenize the input before running the hilbert driven GB algorithm if it is not already homogeneous and dehomogenize the resulting Gröbner basis to obtain one for the original input. This only works if the `base_ring` of the input is not already a graded ring, I am not sure what to do in this case. See https://github.com/oscar-system/Oscar.jl/pull/1821/commits/5b03dcde8bed103b53cc805747488a0d266e229f and https://github.com/oscar-system/Oscar.jl/pull/1821/commits/1fd6300be3232cc8efc4e61b933fc7b0a340be5e for the corresponding changes.

ederc commented on src/Rings/groebner.jl:
```diff
@@ -1221,3 +1248,126 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
+positive characteristic, otherwise compute the Hilbert series for `I`
+modulo a randomly chosen prime. Use the resulting Hilbert series to
+optimize the Gröbner basis computation for `I`
+w.r.t. `target_ordering`.
```
There is no `target_ordering`

ederc commented on src/Rings/groebner.jl:
```diff
@@ -1221,3 +1248,126 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
+
+@doc Markdown.doc"""
+    groebner_basis_hilbert_driven(I::MPolyIdeal,
+                                  destination_ordering::MonomialOrdering;
```
Please rename it to `ordering` and make it a keyword argument to coincide with our naming/argument convention.

RafaelDavidMohr replied:
What should be the default value for `ordering`?

ederc replied:
I would propose that we do not have default value here. For me it is the same as with `fglm`, we do not know the destination ordering of the GB, this is more of a specialized function where we expect the users to know what they want.

ederc commented:
Please add some tests, for example,

```
R,(x,y) = PolynomialRing(QQ, ["x","y"])
I = ideal([x*(x+1), x^2-y^2+(x-2)*y])
standard_basis(I, ordering=lex(R), algorithm=:hilbert)
```

is broken.

RafaelDavidMohr committed:
fixes a coercion bug in `groebner_basis_hilbert_driven`

RafaelDavidMohr committed:
adds some tests for `groebner_basis_hilbert_driven`

RafaelDavidMohr committed:
make ordering a kwarg in `groebner_basis_hilbert_driven`

RafaelDavidMohr committed:
changes `groebner_basis_hilbert_driven` to only take `MPolyRing_dec`

fingolfin commented on src/Rings/groebner.jl:
```diff
@@ -1,7 +1,8 @@
 		groebner_basis_with_transformation_matrix,
-		leading_ideal, syzygy_generators, is_standard_basis, is_groebner_basis
+	leading_ideal, syzygy_generators, is_standard_basis, is_groebner_basis,
+  groebner_basis_hilbert_driven
```
Please don't use tabs (ideally configure your editor to not use tabs in the OSCAR code base -- e.g. by configuring to honor your `.editorconfig` file, see https://editorconfig.org).
```suggestion
		leading_ideal, syzygy_generators, is_standard_basis, is_groebner_basis,
		groebner_basis_hilbert_driven
```

fingolfin commented on src/Rings/groebner.jl:
```diff
@@ -150,6 +153,22 @@ function standard_basis(I::MPolyIdeal; ordering::MonomialOrdering = default_orde
 	elseif algorithm == :fglm
 		_compute_groebner_basis_using_fglm(I, ordering)
+  elseif algorithm == :hilbert
+    if typeof(base_ring(I)) <: MPolyRing_dec
```
Better:
```suggestion
    if base_ring(I) isa MPolyRing_dec
```
Potentially even better: use a helper function with two methods that does this, but only if it can be reused (so don't bother)

fingolfin commented on src/Rings/groebner.jl:
```diff
@@ -249,7 +269,7 @@ julia> leading_coefficient(G[8])
 """
 function groebner_basis(I::MPolyIdeal; ordering::MonomialOrdering = default_ordering(base_ring(I)), complete_reduction::Bool=false,
-	algorithm::Symbol = :buchberger)
+                        algorithm::Symbol = :buchberger)
```
Here you de-tab, thank you :-)

fingolfin commented on src/Rings/groebner.jl:
```diff
@@ -1221,3 +1243,124 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
+                                       ordering::MonomialOrdering,
+                                       complete_reduction::Bool = false) where {P <: MPolyElem_dec}
+  
+  all(p -> is_homogeneous(p), gens(I)) || error("I must be given by generators homogeneous with respect to its underlying ring.")
```
```suggestion
  all(is_homogeneous, gens(I)) || error("I must be given by generators homogeneous with respect to its underlying ring")
```
or even just
```suggestion
  all(is_homogeneous, gens(I)) || error("I must be given by homogeneous generators")
```

RafaelDavidMohr replied:
I wrote the error message in this way because the base ring of `I` may have non-standard grading and wanted to make it completely clear that homogeneity is meant with respect to this base ring. Do you nonetheless prefer your version?

fingolfin replied:
If you prefer to leave the message, that's fine by me -- that's why I made two alternative suggestions. Just apply the first?

fingolfin commented on src/Rings/groebner.jl:
```diff
@@ -1221,3 +1243,124 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
+  if isempty(I.gb) && iszero(characteristic(base_ring(I)))  
+    while true
+      r = rand(2^15:2^16)
+      p = Hecke.next_prime(r)
```
Just wondering, but: is there any particular reason to pick a random `r` each time (and thus potentially reusing the same prime, instead of just doing this:
```suggestion
    p = 32771
    while true
      p = Hecke.next_prime(p)
```

fingolfin commented on src/Rings/groebner.jl:
```diff
@@ -1221,3 +1243,124 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
+        G = groebner_assure(ideal(ModP, I_mod_p_gens), default_ordering(ModP))
+        break
+      catch e
+        if e == ErrorException("Unable to coerce") 
```
Maybe add a comment that explains where this is supposed to be (potentially) thrown? Presumably it is just the conversion/coercion to `ModP`, but not the call to `groebner_assure`? If the `groebner_assure` can also throw, that should be mentioned. If it can't, perhaps move it out of the `try/catch`?

Thing is, I am very slightly concerned about future breakage where the hardcoded  error message changes and we then might have a hard time restoring this, so it's helpful to not what is supposed to throw

Maybe we should add a `CoercionException` or so?

RafaelDavidMohr replied:
I added such a comment, see https://github.com/oscar-system/Oscar.jl/pull/1821/commits/1c944f41fd60c7aeee67eeb3d607ebd763757e83. I'd be happy to also try and add a `CoercionException`.

fingolfin replied:
`CoercionException` is way beyond the scope of this PR, it shouldn't be hold up by it :-)

fingolfin commented on src/Rings/groebner.jl:
```diff
@@ -1221,3 +1243,124 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
+  all(p -> is_homogeneous(p), gens(I)) || error("I must be given by generators homogeneous with respect to its underlying ring.")
+	isa(coefficient_ring(base_ring(I)), AbstractAlgebra.Field) || error("The underlying coefficient ring of I must be a field.")
+  is_global(ordering) || error("Destination ordering must be global.")
+	haskey(I.gb, ordering) && return I.gb[ordering]
```
```suggestion
  haskey(I.gb, ordering) && return I.gb[ordering]
```

fingolfin commented on src/Rings/groebner.jl:
```diff
@@ -1221,3 +1243,124 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
+  singular_assure(G)
+  weights = _extract_weights(base_ring(G))
+  h = Singular.hilbert_series(G.S, weights)
+	singular_assure(I.gens, ordering)
```
Odd formatting follows due to tabs vs spaces

fingolfin commented on src/Rings/groebner.jl:
```diff
@@ -1221,3 +1243,124 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
+  else
+    return ones(Int, nvars(T))
+  end
+end
```
```suggestion
function _extract_weights(T::MPolyRing_dec)
  if !is_z_graded(T)
    error("Ring must be graded by the Integers.")
  end
  return [Int(first(gr_elem.coeff)) for gr_elem in T.d]
end

function _extract_weights(T::MPolyRing)
  return ones(Int, nvars(T))
end
```

fingolfin commented on test/Rings/groebner-test.jl:
```diff
@@ -1,3 +1,4 @@
+using Test: Error
```
Why is this needed? If it isn't, please remove again.

RafaelDavidMohr replied:
It is not, my language server just keeps adding in these import statements if it thinks something is out of scope... I will pay closer attention to it in the future.

fingolfin commented on src/Rings/groebner.jl:
```diff
@@ -1221,3 +1243,124 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
+            push_term!(ctx, base_field(c), m)
+          end
+          I_mod_p_gens[i] = finish(ctx)
+        end
```
I have not tested this, but: can this perhaps be simplified to
```suggestion
        for (i, f) in enumerate(gens(I))
          I_mod_p_gens[i] = map_coefficients(base_field, f; parent=ModP)
        end
```
or even
```suggestion
        I_mod_p_gens = [map_coefficients(base_field, f; parent=ModP) for f in gens(I)]
```

RafaelDavidMohr replied:
It can! Added in https://github.com/oscar-system/Oscar.jl/pull/1821/commits/ea977e7db3ee648e386f4a30508a24ae9ecab226

fingolfin commented on src/Rings/groebner.jl:
```diff
@@ -322,6 +342,8 @@ function f4(
 end
 
+  
+
```
```suggestion
```

fingolfin reviewed (commented):
Thank you for working on this. I have some additional minor feedback

RafaelDavidMohr committed:
remove tabs in `export` statement

Co-authored-by: Max Horn <max@quendi.de>

RafaelDavidMohr committed:
change inheritance check to `isa`

Co-authored-by: Max Horn <max@quendi.de>

RafaelDavidMohr committed:
safer random prime generation

Co-authored-by: Max Horn <max@quendi.de>

RafaelDavidMohr committed:
adds comment for coercion error

RafaelDavidMohr committed:
fixes indent

Co-authored-by: Max Horn <max@quendi.de>

RafaelDavidMohr committed:
Split up `_extract_weights` into two functions

Co-authored-by: Max Horn <max@quendi.de>

RafaelDavidMohr committed:
removes superflous `using` in tests

RafaelDavidMohr committed:
simpler coercion code

RafaelDavidMohr committed:
remove superflous function

RafaelDavidMohr committed:
remove superflous empty line

fingolfin committed:
Update src/Rings/groebner.jl

fingolfin committed:
Avoid an allocation, formatting tweak

fingolfin committed:
tabs to spaces

fingolfin reviewed (approved):
I've pushed two minor commits. Thanks!
